test/conftest.py

```diff
@@ -291,56 +286,186 @@ def pytest_collection_modifyitems(items: List[Any]) -> None:
-                failed_index = index
-                if self.skip:
-                    failed_index += 1
-                break
-
-        # If the previously failed test was not found among the test items,
-        # do not skip any tests.
-        if failed_index is None:
-            self.report_status = "previously run test not found, not skipping."
+        # validate that the cache is correct
+        for item, test in zip(
+            items,
+            self.already_ran + self.to_run,
+        ):
+            assert (
+                item.nodeid == test[0]
+            ), f"Cache and discovered tests do not match, {item.nodeid} != {test[0]}"
+
+        first_test, status = self.to_run[0]
+        first_index = self.get_index(first_test, items)
+        self.prev_run = []
+        deselected = []
+
+        if status == "cont":
+            # We're in the middle of a test run
+            deselected = items[:first_index]
+            items[:] = items[first_index:]
+            self.report_status += f"Continuing from {first_test}"
+            i, test = next(
+                ((i, t) for i, t in enumerate(self.to_run) if t[1] != "cont"),
+                (len(self.to_run), None),
+            )
+            deselected += items[i:]
+            items[:] = items[:i]
+            if test is not None:
+                self.report_status += f", stopping at {test[0]} (exclusive)"
+            self.prev_run = self.to_run[:i]
         else:
-            self.report_status = f"skipping {failed_index} already run items."
-            deselected = items[:failed_index]
-            del items[:failed_index]
-            if self.run_single:
-                self.report_status += f" Running only {items[0].nodeid}"
-                deselected += items[1:]
-                del items[1:]
-            config.hook.pytest_deselected(items=deselected)
+            # Run single test
+            deselected = items[:first_index] + items[first_index + 1 :]
+            items[:] = items[first_index : first_index + 1]
+            self.prev_run = [[first_test, status]]
+            self.report_status += f"Running single test {first_test}"
+
+        self.pytest_previous_status = "no xml"
+        self.save_cache()
+        config.hook.pytest_deselected(items=deselected)
 
     def pytest_report_collectionfinish(self) -> Optional[str]:
         if self.config.getoption("verbose") >= 0 and self.report_status:
             return f"stepcurrent: {self.report_status}"
         return None
 
     def pytest_runtest_protocol(self, item, nextitem) -> None:
-        self.lastrun = item.nodeid
-        self.cache.set(self.directory, self.lastrun)
+        self.ended_at = item.nodeid
+        self.save_cache_active()
+
+    def save_cache(self) -> None:
+        self.cache.set(
+            self.cache_info_init_path,
+            {
+                "prev_run": self.prev_run,
+                "to_run": self.to_run,
+                "already_ran": self.already_ran,
```

Comment:
Could you please also document the format of these cache files somewhere?  It's a bit tricky to figure out exactly what they contain.

Ideal behavior would be to define them as lists of a `typing.NamedTuple`